envoy/server/factory_context.h

```diff
@@ -177,14 +205,36 @@ class ServerFactoryContext : public virtual CommonFactoryContext {
    * @return envoy::config::bootstrap::v3::Bootstrap& the servers bootstrap configuration.
    */
   virtual envoy::config::bootstrap::v3::Bootstrap& bootstrap() PURE;
+
+  /**
+   * @return Http::Context& a reference to the http context.
+   */
+  virtual Http::Context& httpContext() PURE;
+
+  /**
+   * @return whether external healthchecks are currently failed or not.
+   */
+  virtual bool healthCheckFailed() PURE;
+
+  /**
+   * @return ProcessContextOptRef an optional reference to the
+   * process context. Will be unset when running in validation mode.
+   */
+  virtual ProcessContextOptRef processContext() PURE;
+
+  /**
+   * @return the downstream factory context, for downstream filters.
+   */
+  virtual OptRef<DownstreamFactoryContext> downstreamContext() PURE;
```

Comment:
I'm a bit concerned about subtle lifetime issues here with ECDS. An ECDS filter can outlive the parent listener when shared, so a reference here would become stale in a non-obvious way.

Reply:
Can you expand more on how a filter would be shared? It seems strange to me that it would work this way if ultimately the filters need to get information for the listener they are responding on? I don't know much about ECDS.

Assuming this is true I guess this needs to be a weak ptr?

Reply:
It happens when listener 1 opens ECDS stream, then listener 2 shares the stream, then listener 1 is deleted. ECDS filters are dis-associated from listeners in some ways. Filters don't really need much from the parent at runtime, but the factories can capture some listener state if that becomes accessible from root factory context. I think we need to make it more explicit the limited lifetime of those references.

Reply:
well right now, the lifetimes are that of the listener, so pretty much guaranteed for the lifetime of all the code calling it.
given this is only actually used in 2 places (wasm / auth, I guess 3 if you count the sds test) I'm inclined to land as-is with no lifetime changes, and land ECDS lifetime changes separately in a follow up as it's needed.  Would that work for you?

Reply:
Is the idea to have this populated sometimes, and leave it empty the other times when the lifetime is not guaranteed? I'm confused why a server context points to the more specific listener context.

Reply:
likely abandoning in favor of https://github.com/envoyproxy/envoy/pull/22134